slasher/detection/attestations/types/epoch_store.go

```diff
@@ -0,0 +1,125 @@
+package types
+
+import (
+	"errors"
+
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+)
+
+// EpochStore defines an implementation of the slasher data access interface
+// using byte array as data source to extract and put validator spans into.
+type EpochStore struct {
+	spans              []byte
+	highestObservedIdx uint64
+}
+
+// ErrWrongSize appears when attempting to use epoch store byte array with size that
+// is not a multiple of spanner encoded length.
+var ErrWrongSize = errors.New("wrong data length for min max span byte array")
+
+// NewEpochStore initialize epoch store from a byte array
+// returns error if byte length is not a multiple of encoded spanner length.
+func NewEpochStore(spans []byte) (*EpochStore, error) {
+	spansLen := uint64(len(spans))
+	if spansLen%SpannerEncodedLength != 0 {
+		return &EpochStore{}, ErrWrongSize
+	}
+	highestIdx := spansLen / SpannerEncodedLength
+	if highestIdx > 0 {
+		// Minus one here since validators are 0 index.
+		highestIdx--
+	}
+	es := &EpochStore{
+		spans:              spans,
+		highestObservedIdx: highestIdx,
+	}
+	return es, nil
+}
+
+// GetValidatorSpan unmarshal a span from an encoded, flattened array.
+func (es *EpochStore) GetValidatorSpan(idx uint64) (Span, error) {
+	r := Span{}
+	spansLen := uint64(len(es.spans))
+	if spansLen%SpannerEncodedLength != 0 {
+		return r, ErrWrongSize
+	}
+	if idx > es.highestObservedIdx {
+		return Span{}, nil
+	}
+	origLength := uint64(len(es.spans)) / SpannerEncodedLength
+	requestedLength := idx + 1
+	if origLength < requestedLength {
+		return r, nil
+	}
+	cursor := idx * SpannerEncodedLength
+	r.MinSpan = bytesutil.FromBytes2(es.spans[cursor : cursor+2])
+	r.MaxSpan = bytesutil.FromBytes2(es.spans[cursor+2 : cursor+4])
+	sigB := [2]byte{}
+	copy(sigB[:], es.spans[cursor+4:cursor+6])
+	r.SigBytes = sigB
+	r.HasAttested = bytesutil.ToBool(es.spans[cursor+6])
+	return r, nil
+}
+
+// SetValidatorSpan marshal a validator span into an encoded, flattened array.
+func (es *EpochStore) SetValidatorSpan(idx uint64, newSpan Span) (*EpochStore, error) {
+	spansLen := uint64(len(es.spans))
+	if spansLen%SpannerEncodedLength != 0 {
+		return nil, errors.New("wrong data length for min max span byte array")
+	}
+	if es.highestObservedIdx < idx {
```

Comment:
in the original implementation i use global highestObservedIdx in order to avoid reallocations. highestObservedIdx can only grow from epoch to epoch. so we could potentially allocate it to a new epoch or find a better method to allocate once and save a lot of resources

Reply:
Do you think we can keep it this way as it might be safer for now, and later in the future we can look into removing it?

Reply:
ok. lets take it forward. i just changed it for look back removal purposes